node/overseer/src/lib.rs

```diff
@@ -61,70 +61,89 @@
 };
-use futures_timer::Delay;
 use lru::LruCache;
 
 use polkadot_primitives::v1::{Block, BlockId,BlockNumber, Hash, ParachainHost};
 use client::{BlockImportNotification, BlockchainEvents, FinalityNotification};
 use sp_api::{ApiExt, ProvideRuntimeApi};
 
-use polkadot_subsystem::messages::{
+use polkadot_node_network_protocol::{
+	v1 as protocol_v1,
+};
+use polkadot_node_subsystem_types::messages::{
 	CandidateValidationMessage, CandidateBackingMessage,
 	ChainApiMessage, StatementDistributionMessage,
 	AvailabilityDistributionMessage, BitfieldSigningMessage, BitfieldDistributionMessage,
 	ProvisionerMessage, RuntimeApiMessage,
-	AvailabilityStoreMessage, NetworkBridgeMessage, AllMessages, CollationGenerationMessage,
+	AvailabilityStoreMessage, NetworkBridgeMessage, CollationGenerationMessage,
 	CollatorProtocolMessage, AvailabilityRecoveryMessage, ApprovalDistributionMessage,
 	ApprovalVotingMessage, GossipSupportMessage,
+	NetworkBridgeEvent,
+	DisputeParticipationMessage, DisputeCoordinatorMessage, ChainSelectionMessage,
 };
-pub use polkadot_subsystem::{
-	Subsystem, SubsystemContext, SubsystemSender, OverseerSignal, FromOverseer, SubsystemError,
-	SubsystemResult, SpawnedSubsystem, ActiveLeavesUpdate, ActivatedLeaf, DummySubsystem, jaeger,
-	LeafStatus,
+pub use polkadot_node_subsystem_types::{
+	OverseerSignal,
+	errors::{SubsystemResult, SubsystemError,},
+	ActiveLeavesUpdate, ActivatedLeaf, LeafStatus,
+	jaeger,
 };
-use polkadot_node_subsystem_util::{TimeoutExt, metrics::{self, prometheus}, metered, Metronome};
-use polkadot_node_primitives::SpawnNamed;
-use polkadot_procmacro_overseer_subsystems_gen::AllSubsystemsGen;
-
-#[cfg(test)]
-mod tests;
 
-// A capacity of bounded channels inside the overseer.
-const CHANNEL_CAPACITY: usize = 1024;
-// The capacity of signal channels to subsystems.
-const SIGNAL_CHANNEL_CAPACITY: usize = 64;
+/// TODO legacy, to be deleted, left for easier integration
+mod subsystems;
+pub use self::subsystems::AllSubsystems;
 
-// A graceful `Overseer` teardown time delay.
-const STOP_DELAY: u64 = 1;
-// Target for logs.
-const LOG_TARGET: &'static str = "parachain::overseer";
+mod metrics;
+use self::metrics::Metrics;
 
-trait MapSubsystem<T> {
-	type Output;
+use polkadot_node_metrics::{
+	metrics::{
+		prometheus,
+		Metrics as MetricsTrait
+	},
+	Metronome,
+};
+pub use polkadot_overseer_gen::{
+	TimeoutExt,
+	SpawnNamed,
+	Subsystem,
+	SubsystemMeterReadouts,
+	SubsystemMeters,
+	SubsystemIncomingMessages,
+	SubsystemInstance,
+	SubsystemSender,
+	SubsystemContext,
+	overlord,
+	MessagePacket,
+	SignalsReceived,
+	FromOverseer,
+	ToOverseer,
+	MapSubsystem,
+};
+pub use polkadot_overseer_gen as gen;
 
-	fn map_subsystem(&self, sub: T) -> Self::Output;
+/// Boiler plate reduction, `Signal` and `AllMessages` are fixed for one overseer.
+/// So the only variation for external entities is `M`.
+pub trait SubsystemCtx<M> : SubsystemContext<Message = M, Signal = OverseerSignal, AllMessages = AllMessages> {
```

Comment:
Looks indeed useful, but does not seem to be used at all - why?

Reply:
Huh, that should not exist anymore. This also did not proof viable. There are some usage patterns that want to omit `type Message` for usage in multiple subsystem, which can not be achieved with generics.

Reply:
:bulb:

Reply:
Wait, can those usages not just pick a generic `M`?